File: impl/php/src/Mron.php
<?php

declare(strict_types=1);

namespace Makrell\Formats;

final class Mron
{
    public static function parseString(string $source): mixed
    {
        $nodes = MiniMbf::parse($source);
        if ($nodes === []) {
            return null;
        }
        if (count($nodes) === 1) {
            return self::convertNode($nodes[0]);
        }
        if (count($nodes) % 2 !== 0) {
            throw new MakrellFormatException('Illegal number (' . count($nodes) . ') of root level expressions for MRON object.');
        }
        return self::convertPairs($nodes);
    }

    public static function parseFile(string $path): mixed
    {
        $source = @file_get_contents($path);
        if ($source === false) {
            throw new MakrellFormatException('Could not read MRON file: ' . $path);
        }
        return self::parseString($source);
    }

    public static function writeString(mixed $value): string
    {
        return self::writeValue($value);
    }

    private static function convertNode(array $node): mixed
    {
        return match ($node['kind']) {
            'scalar' => self::convertScalar($node['text'], $node['quoted']),
            'square' => array_map(self::convertNode(...), $node['children']),
            'brace' => self::convertPairs($node['children']),
            default => throw new MakrellFormatException('Unsupported MRON node kind: ' . $node['kind']),
        };
    }

    private static function convertPairs(array $nodes): array
    {
        if (count($nodes) % 2 !== 0) {
            throw new MakrellFormatException('Odd pair count in MRON object.');
        }
        $result = [];
        for ($i = 0; $i < count($nodes); $i += 2) {
            $result[(string) self::convertNode($nodes[$i])] = self::convertNode($nodes[$i + 1]);
        }
        return $result;
    }

    private static function convertScalar(string $text, bool $quoted): mixed
    {
        if ($quoted) {
            return $text;
        }
        return match ($text) {
            'null' => null,
            'true' => true,
            'false' => false,
            default => self::parseNumberOrString($text),
        };
    }

    private static function parseNumberOrString(string $text): mixed
    {
        if (preg_match('/^-?\d+$/', $text) === 1) {
            return (int) $text;
        }
        if (preg_match('/^-?\d+\.\d+$/', $text) === 1) {
            return (float) $text;
        }
        return $text;
    }

    private static function writeValue(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_int($value) || is_float($value)) {
            return (string) $value;
        }
        if (is_string($value)) {
            return self::isIdentifierLike($value) ? $value : self::quote($value);
        }
        if (is_array($value)) {
            if (array_is_list($value)) {
                return '[' . implode(' ', array_map(self::writeValue(...), $value)) . ']';
            }
            $parts = [];
            foreach ($value as $key => $item) {
                $parts[] = self::writeValue((string) $key);
                $parts[] = self::writeValue($item);
            }
            return '{ ' . implode(' ', $parts) . ' }';
        }
        throw new MakrellFormatException('Unsupported MRON value for serialisation.');
    }

    private static function isIdentifierLike(string $value): bool
    {
        return preg_match('/^[A-Za-z_][A-Za-z0-9_-]*$/', $value) === 1;
    }

    private static function quote(string $value): string
    {
        return '"' . addcslashes($value, "\\\"") . '"';
    }
}

File: impl/rust/src/mron.rs
use crate::mini_mbf::{parse as parse_nodes, Node};
use crate::MakrellFormatError;
use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone, PartialEq)]
pub enum MronValue {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    String(String),
    Array(Vec<MronValue>),
    Object(BTreeMap<String, MronValue>),
}

pub fn parse_string(source: &str) -> Result<MronValue, MakrellFormatError> {
    let nodes = parse_nodes(source)?;
    if nodes.is_empty() {
        return Ok(MronValue::Null);
    }
    if nodes.len() == 1 {
        return convert_node(&nodes[0]);
    }
    if nodes.len() % 2 != 0 {
        return Err(MakrellFormatError::new(format!(
            "Illegal number ({}) of root level expressions for MRON object.",
            nodes.len()
        )));
    }
    convert_pairs(&nodes)
}

pub fn parse_file(path: impl AsRef<Path>) -> Result<MronValue, MakrellFormatError> {
    let path_ref = path.as_ref();
    let source = fs::read_to_string(path_ref)
        .map_err(|_| MakrellFormatError::new(format!("Could not read MRON file: {}", path_ref.display())))?;
    parse_string(&source)
}

pub fn write_string(value: &MronValue) -> Result<String, MakrellFormatError> {
    Ok(write_value(value))
}

fn convert_node(node: &Node) -> Result<MronValue, MakrellFormatError> {
    match node {
        Node::Scalar { text, quoted } => Ok(convert_scalar(text, *quoted)),
        Node::Square(children) => {
            let mut values = Vec::new();
            for child in children {
                values.push(convert_node(child)?);
            }
            Ok(MronValue::Array(values))
        }
        Node::Brace(children) => convert_pairs(children),
        Node::Paren(_) => Err(MakrellFormatError::new("Parenthesised nodes are not valid MRON values.")),
    }
}

fn convert_pairs(nodes: &[Node]) -> Result<MronValue, MakrellFormatError> {
    if nodes.len() % 2 != 0 {
        return Err(MakrellFormatError::new("Odd pair count in MRON object."));
    }
    let mut result = BTreeMap::new();
    for chunk in nodes.chunks(2) {
        let key = convert_node(&chunk[0])?;
        let value = convert_node(&chunk[1])?;
        result.insert(as_key(key), value);
    }
    Ok(MronValue::Object(result))
}

fn convert_scalar(text: &str, quoted: bool) -> MronValue {
    if quoted {
        return MronValue::String(text.to_string());
    }
    match text {
        "null" => MronValue::Null,
        "true" => MronValue::Bool(true),
        "false" => MronValue::Bool(false),
        _ => {
            if let Ok(value) = text.parse::<i64>() {
                MronValue::Int(value)
            } else if let Ok(value) = text.parse::<f64>() {
                MronValue::Float(value)
            } else {
                MronValue::String(text.to_string())
            }
        }
    }
}

fn as_key(value: MronValue) -> String {
    match value {
        MronValue::String(text) => text,
        MronValue::Null => "null".to_string(),
        MronValue::Bool(value) => value.to_string(),
        MronValue::Int(value) => value.to_string(),
        MronValue::Float(value) => value.to_string(),
        MronValue::Array(_) | MronValue::Object(_) => "[complex]".to_string(),
    }
}

fn write_value(value: &MronValue) -> String {
    match value {
        MronValue::Null => "null".to_string(),
        MronValue::Bool(value) => value.to_string(),
        MronValue::Int(value) => value.to_string(),
        MronValue::Float(value) => value.to_string(),
        MronValue::String(text) => {
            if is_identifier_like(text) {
                text.clone()
            } else {
                quote(text)
            }
        }
        MronValue::Array(items) => {
            let parts: Vec<String> = items.iter().map(write_value).collect();
            format!("[{}]", parts.join(" "))
        }
        MronValue::Object(map) => {
            let mut parts = Vec::new();
            for (key, value) in map {
                parts.push(write_value(&MronValue::String(key.clone())));
                parts.push(write_value(value));
            }
            format!("{{ {} }}", parts.join(" "))
        }
    }
}

fn is_identifier_like(text: &str) -> bool {
    let mut chars = text.chars();
    match chars.next() {
        Some(first) if first.is_ascii_alphabetic() || first == '_' => {}
        _ => return false,
    }
    chars.all(|ch| ch.is_ascii_alphanumeric() || ch == '_')
}

fn quote(text: &str) -> String {
    format!("\"{}\"", text.replace('\\', "\\\\").replace('"', "\\\""))
}
